### ISSUE TITLE
LM-ResNets/LM ResNexts

### ISSUE THREAD

**Issue by pGit1** [titu1994/Keras-Classification-Models#9]
This paper on your radar? Any idea how to implement something like this?

1. https://arxiv.org/pdf/1710.10121.pdf

**Comment by titu1994** [comment]
Seems easy enough to implement by extending the Add() layer or even just duplicating the code and then adding the weights to be trained. I won't be implementing this paper though because I'm busy in my final semester. Maybe if I get some free time on some weekend.

**Comment by titu1994** [comment]
Also, ResNeXt versions can't be built in Keras / TF. I don't think I'll be making it after all, probably just the layer so that it can be used as required.

**Comment by pGit1** [comment]
Sounds good @titu1994.

If I come across anything else I'll keep on your radar. Not sure how you can parse the idea that fast but its impressive. For the life of me I cant seem to figure out what weights are actually being optimized here.

**Comment by titu1994** [comment]
The weights being optimized are k_n for the input u_n-1 from the first residual block and u_n (the residual) and f(u_n) from the second residual block. When adding a residual connection in Keras you generally do something like `add([input, x])`. Here, u_n is input and x is f(u_n). This is repeated after each conv blocks. 

However, this paper is also going to learn a scaling factor between the u_n and u_n-1 and then add the activations from f(u_n) to this scaled residual connection after every two residual blocks. The first block will have the ordinary residual connection, whereas the second one will have the scaling. It's quite ingenious. So this will infact take 3 inputs and not just 2 for the second block.

**Comment by titu1994** [comment]
On the topic of papers being difficult to read and implement, this one is quite a challenge - https://arxiv.org/abs/1802.01808.

But I will probably try to implement this at least to some level because it seems extremely useful.

**Comment by pGit1** [comment]
Interesting. Tell me how to understand things like this so I can stop
bugging you all the time! :D

**Comment by pGit1** [comment]
Is this on your radar? https://arxiv.org/pdf/1801.05895.pdf

For the life of me I am having a hard time with these two images:

![image](https://user-images.githubusercontent.com/13975114/36057309-7eceb09e-0dda-11e8-87cf-0f29f58afda5.png)

and this one

![image](https://user-images.githubusercontent.com/13975114/36057326-acd5ec96-0dda-11e8-878e-e0364101777e.png)

I guess it has something to do with the "offset" but the processing unit on the far right is connected to the second from the left, not the very first from the left in the top image. (i.e. the i-8th unit ). Either way I like this idea of sparse connectivity. Seems like a good one. This is straightforward to implement conceptually but there are details about the 1x1 layers and others I have questions about.

**Comment by titu1994** [comment]
There are no papers in my "radar" atm, because I am swamped with coursework. However, very interesting paper. To be sure, it should be quite direct modification of my current DenseNet repo to incorporate SparseNet connections. The only function to modify would be the dense block really, though that has to be done somewhat carefully. You can give it a shot yourself actually.  

As to the diagram, focus on the second diagram. Create all the required nodes without joining the tensors. Then work backwards from the last tensor and set it's input connection to be as shown in the diagram. Of course, this is for mental image. For Tensorflow or Keras you would need to go forward and decide ahead of time which node connects to which node.

**Comment by pGit1** [comment]
I did this in my own mock implementation but the mental image you describe above conflicts with the first image (which the authors claim is correct). If I work backwards (as you suggest I've done so a million times) I dont see how the two diagrams are equivalent. Working backwards in the top diagram lands me on node corresponding i-7 node in the second diagram. Is this not correct? Losing my mind here. Sorry for bugging you with papers and I understand about your coursework. 

Do you want me to still me to send you papers of interest when I come across them?

**Comment by titu1994** [comment]
Yeah send away please. I am not up to date on the latest and greatest in the field for the past month or so, so I need to catch up after I'm done with the remaining assignments.

Also, I see your point. I made a mistake with the backward iteration. Interesting. Why would they plot to contradicting diagrams is beyond me. Consider this. Node 0 at the leftmost is the input to the dense block. If we consider the forward chaining of inputs, we could determine a few rules.

Do you have a gist of your implementation somewhere? I'm gonna try implementing this tonight. There goes my friday lol

**Comment by pGit1** [comment]
No I don't have a gist but here is what I have. It sucks so dont make fun of me...I implemented what I thought was going on according to the first diagram that they swore was correct...also I have no idea how to architect the models they describe and match their depth and parmeters:

![image](https://user-images.githubusercontent.com/13975114/36058396-d30e9220-0dec-11e8-979d-948c20365885.png)


`

    def conv_block(x,filters=32,sublocks=2,pad='same',kernel_hw=3,decay=1e-4,stride=1):
        for i in range(sublocks):
            x = layers.Conv2D(filters,kernel_hw,padding=pad,kernel_initializer='he_normal',
                              strides=stride)(x)
            x = layers.BatchNormalization()(x)
            x = layers.Activation('relu')(x)

    def sparse_connect(x,fil=36):
        #Input Block
        x1 = conv_block(x,filters=fil,sublocks=1)

        x2 = conv_block(x1,filters=fil,sublocks=1)

        conn = layers.concatenate([x2,x1])
        x3 = conv_block(conn,filters=fil,sublocks=1)

        conn = layers.concatenate([x3,x2])
        x4 = conv_block(conn,filters=fil,sublocks=1)

        conn = layers.concatenate([x4,x3,x1])
        x5 = conv_block(conn,filters=fil,sublocks=1)

        conn = layers.concatenate([x5,x4,x2])
        x6 = conv_block(conn,filters=fil,sublocks=1)

        conn = layers.concatenate([x6,x5,x3])
        x7 = conv_block(conn,filters=fil,sublocks=1)

        conn = layers.concatenate([x7,x6,x4,x1]) #Seems wrong to connect to x1 but authors say its right
        x8 = conv_block(conn,filters=fil,sublocks=1)

        conn = layers.concatenate([x8,x7,x5,x2]) #Seems wrong to connect to x2 but authors say its right
        x9 = conv_block(conn,filters=fil,sublocks=1) 
        return x9

    lr = .01
    def create_model(): #NO IDEA what I'm doing here in terms of replicating their models
        inp = layers.Input(shape=(32,32,3))
        
        #Block 1
        x = sparse_connect(inp,fil=36)
        x = conv_block(x,filters=32,sublocks=1,kernel_hw=1)
        #Transition
        x = layers.AveragePooling2D()(x)
        #Block 2
        x = sparse_connect(x,fil=36)
        x = conv_block(x,filters=64,sublocks=1,kernel_hw=1)
        #Transition
        x = layers.AveragePooling2D()(x)
        #Block3
        x = sparse_connect(x,fil=36)

        #Out
        x = layers.GlobalAvgPool2D()(x)
        out = layers.Dense(10,activation='softmax')(x)
        
        model = models.Model(inputs=inp,outputs=out)
`


EDIT: Sorry about your Friday Night!!

**Comment by titu1994** [comment]
Hmm interesting. I am trying to go by what they write in the second diagram : 
```python
import numpy as np

data = [0]

print("Node: 1 (root)", data)
print()

for i in range(8):
    log = np.round(np.log2(len(data))).astype(int)
    print("Node:", i + 2, "Inputs available:", data, "Log:", log)

    parents = []
    for p in range(0, log + 1):
        index = 2 ** p

        if index <= len(data):
            ip = data[-index]

            print("Adding parent at index : ", -index)
            parents.append(ip)

    print("Parents (%d):" % (i + 2), parents)
    data.append(i + 1)

    print()
```

And getting the output:
```python
Node: 1 (root) [0]

Node: 2 Inputs available: [0] Log: 0
Adding parent at index :  -1
Parents (2): [0]

Node: 3 Inputs available: [0, 1] Log: 1
Adding parent at index :  -1
Adding parent at index :  -2
Parents (3): [1, 0]

Node: 4 Inputs available: [0, 1, 2] Log: 2
Adding parent at index :  -1
Adding parent at index :  -2
Parents (4): [2, 1]

Node: 5 Inputs available: [0, 1, 2, 3] Log: 2
Adding parent at index :  -1
Adding parent at index :  -2
Adding parent at index :  -4
Parents (5): [3, 2, 0]

Node: 6 Inputs available: [0, 1, 2, 3, 4] Log: 2
Adding parent at index :  -1
Adding parent at index :  -2
Adding parent at index :  -4
Parents (6): [4, 3, 1]

Node: 7 Inputs available: [0, 1, 2, 3, 4, 5] Log: 3
Adding parent at index :  -1
Adding parent at index :  -2
Adding parent at index :  -4
Parents (7): [5, 4, 2]

Node: 8 Inputs available: [0, 1, 2, 3, 4, 5, 6] Log: 3
Adding parent at index :  -1
Adding parent at index :  -2
Adding parent at index :  -4
Parents (8): [6, 5, 3]

Node: 9 Inputs available: [0, 1, 2, 3, 4, 5, 6, 7] Log: 3
Adding parent at index :  -1
Adding parent at index :  -2
Adding parent at index :  -4
Adding parent at index :  -8
Parents (9): [7, 6, 4, 0]
```

**Comment by titu1994** [comment]
You'll notice, the final layer's parent (layer 9) has [0] at the end rather than [1] which is in line with how we would ordinarily word if we followed their second diagram to a T.

However, by ploping this into my OG densenet implementation, I can only get 0.71 M parameters for SparseNet 40-24. Where am I missing parameters is a difficult question.. I think ill give up on this for tonight.

**Comment by titu1994** [comment]
Lol, a little bit of Googling the primary author leads me to this : https://github.com/Lyken17/SparseNet/blob/master/src/denselink.py

This should make it a lot clearer what I need to do.

**Comment by titu1994** [comment]
https://github.com/titu1994/keras-SparseNet

There we go. The only issue is a small decrease in the number of parameters compared to the paper. Since I dont have PyTorch, I cant verify the official repos param count.

**Comment by pGit1** [comment]
I still have a few questions about this. If I wanted to use **my** implementation above what would I change to SparseNet-BC (k=36). I want to try and understand this in my own way. The Pytorch code doesn't help much. I pretty much read every line. Aside from following the second graph and not the first, what would I have to change in my code to get SparseNet-BC (k=36)? Also could you test on CIFAR-10 to see if you replicate their results?

Thanks for your help but for once I wanted to implement something in a way that made sense in my head before moving on to a more sophisticated solution (i.e. your implementation). I'm sooooo frustrated right now to say the least...

**Comment by pGit1** [comment]
In your implementation no idea what depth, nb_dense_block, gorwth_rate, reduction is. Or how you were able to gather all this from their paper....

Why arent depth and nb_dense_block the same? Doesn't one affect the other? Granted Ive been speed reading your code. I should be able to get your results with my long way I feel.

THANK YOU for this implementation though. Im testing it now.

**Comment by pGit1** [comment]
@titu1994 

Dude this implementation is really hard in my way. How on earth do I write a function that knows how to do the connectivity scheme programatically? I see your expeonential fetch code but I dont see how you create dense blocks with it.

**Comment by titu1994** [comment]
Of course. Implementing something yourself from scratch is a very valuable learning experience. 

Try seeing the parts of your code which relate to the functions in my code. You have a conv block which is the same so that's good. You have a sparse connect which is hard coded. Try thinking in terms of exponential loops. My original implementation of this loop was a backwards grouping, the current code is most straightforward and forward grouping. 

The exponential index simply tries to get the closest power of 2 smaller than the max of the input. 

BC mode will require modification of your conv block to insert a bottleneck 1x1 conv layer with 4x the channel and so on. 

Also, the initial filters will also not be doubled in case of the sparsenets, unlike DenseNets. 

Also, your transition block will keep track of the compression factor as well. 

It's small things really. 

Nb dense block is default to 3, for 3 blocks with 2 pooling layers. It's configurable since ImageNet will require 5 blocks with 4 pooling layers. 

Growth rate is to follow the paper as to how many filters each new block in the sparse layer will add. 

Reduction is the inverse of compression, the parameter C in BC mode. 

Depth and nb dense block are not the same for convenience. If you want a standard x layers same across each block, use depth. If you want different number of layers in each block, pass this as a list. It will override depth at this point. It's just convenience of design and adaptability to requirements.

**Comment by pGit1** [comment]
That @titu1994. Great answer! Now I've run a test based on my hard coded way and I cannot replicate their accuracy results. If you have GPU access could you run a test and and see if you can replicate their CIFAR-10 results?
